gtsfm/three_view_estimator.py

```diff
@@ -0,0 +1,104 @@
+"""Estimator which operates on a triplet of images to compute relative pose and verified indices.
+
+Authors: Ayush Baid, John Lambert
+"""
+import logging
+
+import dask
+import numpy as np
+from dask.delayed import Delayed
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.frontend.matcher.matcher_base import MatcherBase
+from gtsfm.frontend.verifier.verifier_base import VerifierBase
+
+logger = logger_utils.get_logger()
+
+
+class ThreeViewEstimator:
+    """Wrapper for running three-view relative pose estimation on image triplets in the dataset."""
+
+    def __init__(self, matcher: MatcherBase, verifier: VerifierBase, corr_metric_dist_threshold: float) -> None:
+        """Initializes the three-view estimator from matcher and verifier.
+
+        Args:
+            matcher: matcher to use for image pairs.
+            verifier: verifier to use.
+            corr_metric_dist_threshold: distance threshold for marking a correspondence pair as inlier. 
+        """
+        self._matcher = matcher
+        self._verifier = verifier
+        self._corr_metric_dist_threshold = corr_metric_dist_threshold
+
+    @classmethod
+    def match_triplet_using_pairwise_matches(
+        cls, match_indices_i1i2: np.ndarray, match_indices_i2i3: np.ndarray, match_indices_i1i3: np.ndarray
+    ) -> np.ndarray:
+        """Form correspondences across triplets by using pairwise matches.
+
+        Note: this logic may not apply to all types of matchers (e.g. ones with ratio test).
+
+        Args:
+            match_indices_i1i2: indices of matches between image i1 and i2, of shape (N1, 2).
+            match_indices_i2i3: indices of matches between image i2 and i3, of shape (N2, 2).
+            match_indices_i1i3: indices of matches between image i1 and i3, of shape (N3, 2).
+
+        Returns:
+            indices of matches between the triplet (i1, i2, i3), of shape (N, 3), where N <= min(N1, N2, N3).
+        """
+        i1_to_i2_dict = {}
+        # TODO: remove this dictionary, not needed between i1 and i2
+        for idx_i1, idx_i2 in match_indices_i1i2:
+            i1_to_i2_dict[idx_i1] = idx_i2
+
+        i2_to_i3_dict = {}
+        for idx_i2, idx_i3 in match_indices_i2i3:
+            i2_to_i3_dict[idx_i2] = idx_i3
+
+        i1_to_i3_dict = {}
+        for idx_i1, idx_i3 in match_indices_i1i3:
+            i1_to_i3_dict[idx_i1] = idx_i3
+
+        # for the triplet matches, chaining i1->i2 with i2->i3 should be present in i1->i3
```

Comment:
seems like we are putting 3-cycle consistency code here. is it how Theia implements their cycle consistency here? 


(https://github.com/sweeneychris/TheiaSfM/blob/master/src/theia/sfm/filter_view_graph_cycles_by_rotation.h or https://github.com/sweeneychris/TheiaSfM/blob/master/src/theia/sfm/filter_view_pairs_from_relative_translation.h

Reply:
That's a different type of cycle consistency. They check the cumulative translation/rotation errors between triplets to throw away cameras.

Here, we do a triplet matching of a correspondence. If point x2 in image i2 is matched with point x1 in image i1, and x3 in image i3, then x1 has to match to x3 too.

Reply:
I see. Do you mind adding this distinction (and references) to the docstring?